### PR TITLE
Big stuffs

### PR DESCRIPTION
**Pull request by ev0x**
- Fixed str formatting
- Updated CUSTOM_EMOJI_STATS_SERVER_SQL and CUSTOM_EMOJI_STATS_GLOBAL_SQL to use count(gm.emoji_id) as it is faster to count off the primary key.
- Updated CUSTOM_EMOJI_STATS_SERVER_SQL and CUSTOM_EMOJI_STATS_GLOBAL_SQL to exclude deleted messages
- Updated CUSTOM_EMOJI_STATS_SERVER_SQL and CUSTOM_EMOJI_STATS_GLOBAL_SQL with `AND cardinality(m.emojis) > 0` to exclude empty arrays
- Rewrote tracking command using MessageTable()
- Updated roles command with help of Xenthys including counts, ordered roles, and patten matching to search for specific roles eg `!roles rolename`
- Fix for this say error https://trello.com/c/ifZPWawQ/28-say-check-for-user-in-guild
- Changed backups default group level from MOD to ADMIN
- Changed archive default group level from MOD to ADMIN
- Added guild to Command Error and Message (command) details to give more context
- Updated time formatting to `strftime("%d-%b-%y @ %H:%M")`
- Added mass bans, mass unmutes, mass mutes and mass unbans
- Handled context errors in report command
- Just continue on error in get_dominant_colors_* as people probably want the info :)
- Moved global_admin and global_bypass into own function in core
- Added emitter for custcmd and tags into core
- Custcmd and tags will be emitted in core and push into the redis queue 
- Handlers for custcmds and tags including the level locking etc have been moved into a new plugin called msghandler
- msghandler now handles DM's not core anymore
- Moved all the commands out of core into a new plugin called global. This also allows us to handle writing code without restarting everything if we need to touch core (Which we hopefully don't anymore)
- Added adv info command for global admins. It is not guild bound and shows extra information
- Added adv inf search for global admins. It allows for GA to search any infraction/note from any guild
- Added adv inf info for global admins. It allows to view any infraction/note from any guild
- Set disco Client max_reconnects to 0
- search command with 1 result will no longer leak if the user is not in the guild you run it in
- set search command default level to mod
- added server owner to server command
- set info command default level to mod
- info command will still show basic info of members not in any of the guilds
- info command includes count of known guilds but not names. The names are included in the adv info (global admin only)
- info command now identifies a global admin and puts this information under the Heimdallr Staff heading
- 3 new plugins for config.yaml need to be defined
```
rowboat.plugins.global
rowboat.plugins.msghandler
```

### CODE REVIEW

File: rowboat/plugins/global.py [hunk 1]
```diff
@@ -0,0 +1,871 @@
...
+                humanize.naturaldelta(datetime.utcnow() - nmsg),
+                nmsg.strftime("%d-%b-%y @ %H:%M"),
+            ))
+            content.append('First Message: {} ago ({})'.format(
+                humanize.naturaldelta(datetime.utcnow() - omsg),
+                omsg.strftime("%d-%b-%y @ %H:%M"),
+            ))
+
+        for inf in infractions:
+            print("HERE")
```
Comment (TobiTenno) [inline rowboat/plugins/global.py]: remove prints, uncomment

File: rowboat/plugins/infractions.py [hunk 2]
```diff
@@ -909,24 +1107,36 @@ def kick(self, event, user, reason=None):
                 u=member.user,
             ))
         else:
-            raise CommandFail('invalid user')
+            raise CommandFail('Invalid user')
 
     @Plugin.command('mkick', parser=True, level=CommandLevels.MOD)
-    @Plugin.parser.add_argument('users', type=long, nargs='+')
+    @Plugin.parser.add_argument('users', type=str, nargs='+')
```
Comment (TobiTenno) [inline rowboat/plugins/infractions.py]: this one had worked before, any particular reason for the change?

File: rowboat/plugins/infractions.py [hunk 3]
```diff
@@ -1053,8 +1327,15 @@ def warn(self, event, user, reason=None):
 
     @Plugin.command('delete', '<infraction:int>', group='infractions', level=CommandLevels.ADMIN)
     def infraction_delete(self, event, infraction):
-        # Written by the wonderful coders over on https://github.com/OGNova/airplane, specifically Tiemen
-        query = "DELETE FROM infractions WHERE id=%s and guild_id = %s and type != 9 RETURNING *;"
+        try:
+            infraction = Infraction.select(Infraction).where(
+                (Infraction.guild_id == event.guild.id) &
+                (Infraction.id == infraction)
+            ).get()
```
Comment (TobiTenno) [inline rowboat/plugins/infractions.py]: did you mean to drop the notes type?

File: rowboat/plugins/msghandler.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                    return
+
+            channel.send_message(u'{}'.format(
+                tag.content
+            ))
+            return
+        return
+
+    @Plugin.listen('MessageCreate')
+    def on_message_create(self, event):
```
Comment (TobiTenno) [inline rowboat/plugins/msghandler.py]: would it matter if we renamed it `on_dm_message_create` so it's not confusing?

File: rowboat/plugins/utilities.py [hunk 5]
```diff
@@ -299,15 +315,15 @@ def seen(self, event, user):
...
             ).order_by(Message.timestamp.desc()).limit(1).get()
         except Message.DoesNotExist:
-            return event.msg.reply(u"I've never seen {}".format(user))
+            return event.msg.reply(u"I've never seen {}".format(user)).after(5).delete()
 
         event.msg.reply(u'I last saw {} {} ago (at {})'.format(
             user,
             humanize.naturaldelta(datetime.utcnow() - msg.timestamp),
-            msg.timestamp
+            msg.timestamp.strftime("%d-%b-%y @ %H:%M")
```
Comment (TobiTenno) [inline rowboat/plugins/utilities.py]: consider adding `GMT` or `UTC` so people don't think it's localized?

Comment (TobiTenno) [reply rowboat/plugins/utilities.py]: also consider moving timestamp format to constants so we don't reuse it over and over?

File: rowboat/plugins/utilities.py [hunk 6]
```diff
@@ -420,27 +462,29 @@ def info(self, event, user):
         created_dt = to_datetime(user.id)
         content.append('Created: {} ago ({})'.format(
             humanize.naturaldelta(datetime.utcnow() - created_dt),
-            created_dt.isoformat()
+            created_dt.strftime("%d-%b-%y @ %H:%M")
         ))
 
+        if rdb.sismember('global_admins', user.id):
+            content.append(u'\n**\u276F Heimdallr Staff**')
```
Comment (TobiTenno) [inline rowboat/plugins/utilities.py]: There should be a `ROWBOAT_NAME`  param someplace

File: frontend/src/components/guild_config_history.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    };
+
+    this.setconfigselected = this.setconfigselected.bind(this);
+  }
+
+  componentWillMount() {
+    globalState.getGuild(this.props.params.gid).then((guild) => {
+      globalState.currentGuild = guild;
+
+      guild.getConfig(true).then((config) => {
```
[the remark this answers is not in the document]
Comment (ev0x) [reply frontend/src/components/guild_config_history.js]: I think I did this correct now

File: frontend/src/components/guild_config_history.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    this.messageTimer = setTimeout(() => {
+      this.setState({
+        message: null,
+      });
+      this.messageTimer = null;
+    }, 5000);
+  }
+
+  setconfigselected(e) {
+    var index = e.target.value
```
[the remark this answers is not in the document]
Comment (ev0x) [reply frontend/src/components/guild_config_history.js]: fixed

File: frontend/src/components/guild_stats.js [hunk 9]
```diff
@@ -1,10 +1,88 @@
...
 export default class GuildStats extends Component {
+  constructor() {
+    super();
+
+    this.state = {
+      message: null,
+      guild: null,
+      mstats: {},
+    };
+
```
[the remark this answers is not in the document]
Comment (ev0x) [reply frontend/src/components/guild_stats.js]: fixed

File: frontend/src/index.html [hunk 10]
```diff
@@ -5,6 +5,7 @@
 
     <link href="/static/css/bootstrap.min.css" rel="stylesheet">
     <link href="/static/css/style.css" rel="stylesheet">
+		<link href="/static/css/diff.css" rel="stylesheet">
```
[the remark this answers is not in the document]
Comment (ev0x) [reply frontend/src/index.html]: fixed

File: rowboat/plugins/global.py (hunk 1 above)
Comment (ev0x) [reply rowboat/plugins/global.py]: fixed

File: rowboat/plugins/infractions.py (hunk 2 above)
Comment (ev0x) [reply rowboat/plugins/infractions.py]: Just inline with all the other mass ones and we can handle the types etc in the function itself

File: rowboat/plugins/infractions.py (hunk 3 above)
Comment (ev0x) [reply rowboat/plugins/infractions.py]: oops.. Fixed this and also fixed the notes one as the query whole function was a bit broken

File: rowboat/plugins/msghandler.py (hunk 4 above)
Comment (ev0x) [reply rowboat/plugins/msghandler.py]: fixed

File: rowboat/plugins/utilities.py (hunk 6 above)
Comment (ev0x) [reply rowboat/plugins/utilities.py]: fixed

File: rowboat/plugins/utilities.py (hunk 5 above)
Comment (ev0x) [reply rowboat/plugins/utilities.py]: At the moment I just replaced everything from isoformat to strftime to clean up the formatting

File: rowboat/plugins/infractions.py (hunk 2 above)
Comment (TobiTenno) [reply rowboat/plugins/infractions.py]: 👍

**Comment by TobiTenno** [comment]
```

Traceback (most recent call last):,
  File "/usr/local/lib/python2.7/runpy.py", line 174, in _run_module_as_main,
    "__main__", fname, loader, pkg_name),
  File "/usr/local/lib/python2.7/runpy.py", line 72, in _run_code,
    exec code in run_globals,
  File "/src/disco-py/disco/cli.py", line 110, in <module>,
    disco_main(True),
  File "/src/disco-py/disco/cli.py", line 101, in disco_main,
    bot = Bot(client, bot_config),
  File "/src/disco-py/disco/bot/bot.py", line 186, in __init__,
    self.add_plugin_module(plugin_mod),
  File "/src/disco-py/disco/bot/bot.py", line 507, in add_plugin_module,
    self.add_plugin(entry, config),
  File "rowboat/plugins/core.py", line 75, in our_add_plugin,
    Bot.add_plugin(self.bot, inst, *args, **kwargs),
  File "/src/disco-py/disco/bot/bot.py", line 456, in add_plugin,
    self.plugins[inst.__class__.__name__].load(ctx or {}),
  File "rowboat/plugins/core_commands.py", line 14, in load,
    self.startup = ctx.get('startup', datetime.utcnow()),
NameError: global name 'datetime' is not defined
```
error i'm getting

**Comment by ev0x** [comment]
[Don't forget to remove core_commands from config.yaml re this](https://github.com/TobiTenno/rowboat/pull/49/commits/4118b62241b0528f933e749b9fcd4f63e573e912)

**Comment by ev0x** [comment]
re 592249e
- Connect to bot container (docker-compose exec bot /bin/bash)
- Update requirements and wipe (w) the disco clone
- Restart bot container

**Comment by ev0x** [comment]
f85c371 is in ref to report command embed field for context exceeding 1024 bytes in come cases

**Pull request merged**